test/Pipeline.hs

```diff
@@ -0,0 +1,196 @@
+  Usage.T ->
+  HR.Term PrimTy PrimVal ->
+  Typed.Globals PrimTy PrimVal ->
+  IO (Either String EmptyInstr)
+toMichelson term usage ty globals = do
+  (res, _) <- exec (P.coreToMichelson term usage ty) michelson globals
+  pure $ case res of
+    Right r ->
+      case r of
+        Right e -> Right e
+        Left err -> Left (show err)
+    Left err -> Left (show err)
+
+toMichelsonContract ::
+  HR.Term PrimTy PrimVal ->
+  Usage.T ->
+  HR.Term PrimTy PrimVal ->
+  Typed.Globals PrimTy PrimVal ->
+  IO (Either String Text)
+toMichelsonContract term usage ty globals = do
+  (res, _) <- exec (P.coreToMichelsonContract term usage ty) michelson globals
+  pure $ case res of
+    Right r ->
+      case r of
+        Right e -> Right (M.typedContractToSource $ snd e)
+        Left err -> Left (show err)
+    Left err -> Left (show err)
+
+tests :: [T.TestTree]
+tests =
+  [ test_constant,
+    test_erased_function,
+    test_real_function_apply,
+    test_partial_erase
+  ]
+
+test_constant :: T.TestTree
+test_constant =
+  shouldCompileTo
+    "constant"
+    (twoTerm, Usage.Omega, intTy)
+    emptyGlobals
+    (EmptyInstr (MT.Seq (MT.Nested (MT.PUSH (MT.VInt 2))) MT.Nop))
+
+test_erased_function :: T.TestTree
+test_erased_function =
+  shouldCompileTo
+    "erased function"
+    (erasedLamTerm, Usage.Omega, erasedLamTy)
+    emptyGlobals
+    (EmptyInstr (MT.Seq (MT.Nested (MT.PUSH (MT.VInt 2))) MT.Nop))
+
+test_real_function_apply :: T.TestTree
+test_real_function_apply =
+  shouldCompileTo
+    "real function with application"
+    (appLam, Usage.Omega, intTy)
+    emptyGlobals
+    (EmptyInstr (MT.Seq (MT.Nested (MT.PUSH (MT.VInt 5))) MT.Nop))
+
+test_partial_erase :: T.TestTree
+test_partial_erase =
+  shouldCompileTo
+    "real function with partial erase"
+    (appLam2, Usage.Omega, intTy)
+    emptyGlobals
+    (EmptyInstr (MT.Seq (MT.Nested (MT.PUSH (MT.VInt 12))) MT.Nop))
+
+twoTerm :: HR.Term PrimTy PrimVal
+twoTerm = HR.Elim (HR.Prim (Constant (M.ValueInt 2)))
+
+erasedLamTerm :: HR.Term PrimTy PrimVal
+erasedLamTerm = HR.Lam "x" (HR.Elim (HR.Prim (Constant (M.ValueInt 2))))
+
+erasedLamTy :: HR.Term PrimTy PrimVal
+erasedLamTy = HR.Pi (Usage.SNat 0) "x" intTy intTy
+
+appLam :: HR.Term PrimTy PrimVal
+appLam = HR.Elim (HR.App (HR.App (HR.Ann (Usage.SNat 1) lamTerm lamTy 0) (HR.Elim (HR.Prim (Constant (M.ValueInt 2))))) (HR.Elim (HR.Prim (Constant (M.ValueInt 3)))))
+
+lamTerm :: HR.Term PrimTy PrimVal
+lamTerm = HR.Lam "x" (HR.Lam "y" (HR.Elim (HR.App (HR.App (HR.Prim AddI) (HR.Elim (HR.Var "x"))) (HR.Elim (HR.Var "y")))))
+
+appLam2 :: HR.Term PrimTy PrimVal
+appLam2 = HR.Elim (HR.App (HR.App (HR.Ann (Usage.SNat 1) lamTerm2 lamTy2 0) (HR.Elim (HR.Prim (Constant (M.ValueInt 2))))) (HR.Elim (HR.Prim (Constant (M.ValueInt 3)))))
+
+lamTerm2 :: HR.Term PrimTy PrimVal
+lamTerm2 = HR.Lam "x" (HR.Lam "y" (HR.Elim (HR.App (HR.App (HR.Prim AddI) (HR.Elim (HR.Var "x"))) (HR.Elim (HR.Prim (Constant (M.ValueInt 10)))))))
+
```

Comment:
Would be nice if we could eventually get these to be smaller lines using `$` or `|>` to avoid extra lines with just ()'s

Reply:
I suppose, but mostly I think it makes sense to wait for quasiquotation and then write the tests that way?